### PR TITLE
feat(telemetry): improve memory consumption

### PR DESCRIPTION
**Pull request by StarpTech**
## Motivation and Context

Based on the included benchmark. The PR reduces memory pressure by many magnitudes. 

CleanShot 2024-11-04 at 00 01 05@2x

Other fixes:

- Host parsing issue with `localhost`
- Status code attribute wasn't attached in all cases

## Checklist

- [ ] I have discussed my proposed changes in an issue and have received approval to proceed.
- [ ] I have followed the coding standards of the project.
- [ ] Tests or benchmarks have been added or updated.
- [ ] Documentation has been updated on [https://github.com/wundergraph/cosmo-docs](https://github.com/wundergraph/cosmo-docs).
- [ ] I have read the [Contributors Guide](https://github.com/wundergraph/cosmo/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -32,8 +32,8 @@
     "dist"
   ],
   "engines": {
-    "node": ">=20.0.0",
-    "pnpm": "8"
+    "node": ">=22.11.0",
```
Comment (df-wg) [inline package.json:35]: Given this change, can you also update https://github.com/wundergraph/cosmo/blob/af0ae3ba2d0afb4a637d101e93e06644cbafb3c4/CONTRIBUTING.md?plain=1#L13-L14 to reflect the new versions?

File: router/pkg/metric/metric_store.go [hunk 2]
```diff
@@ -74,28 +74,34 @@ type (
...
+		MeasureRequestSize(ctx context.Context, contentLength int64, opts ...otelmetric.AddOption)
+		MeasureResponseSize(ctx context.Context, size int64, opts ...otelmetric.AddOption)
+		MeasureLatency(ctx context.Context, latency float64, opts ...otelmetric.RecordOption)
+		MeasureRequestError(ctx context.Context, opts ...otelmetric.AddOption)
+		MeasureOperationPlanningTime(ctx context.Context, planningTime float64, opts ...otelmetric.RecordOption)
 		Flush(ctx context.Context) error
 	}
 
 	Store interface {
+		MeasureInFlight(ctx context.Context, sliceAttr, attr []attribute.KeyValue) func()
```
Comment (df-wg) [inline router/pkg/metric/metric_store.go:96]: Can you add godocs here for how the interfaces are used, and how they are coupled?

**Review by df-wg — CHANGES_REQUESTED** [review]
Overall LGTM, but two docs changes that I think should be added before it is merged

File: package.json (hunk 1 above)
Comment (StarpTech) [reply package.json:35]: Good catch

File: router/pkg/metric/metric_store.go (hunk 2 above)
Comment (StarpTech) [reply router/pkg/metric/metric_store.go:100]: Done

**Review by df-wg — APPROVED** [review]

**Pull request merged**